src/scale.jl

```diff
@@ -575,26 +575,14 @@ element_aesthetics(::ContinuousColorScale) = [:color]
 
 function apply_scale(scale::ContinuousColorScale,
                      aess::Vector{Gadfly.Aesthetics}, datas::Gadfly.Data...)
-    cmin, cmax = Inf, -Inf
-    for data in datas
-        data.color === nothing && continue
-
-        for c in data.color
-            c === NA && continue
-
-            c = convert(Float64, c)
-            if c < cmin
-                cmin = c
-            end
-
-            if c > cmax
-                cmax = c
-            end
-        end
+    cdata = reduce(vcat, [], [i.color for i in datas if i.color != nothing])
+    filter!(!isna, cdata)
+    if !isempty(cdata)
+      cmin, cmax = extrema(cdata)
+    else
+        return
     end
 
-    (cmin == Inf || cmax == -Inf) && return
-
```

Comment:
do the changes above do the exact same thing as before?  or do they have a material bearing on making Unitful work?  if the former, have you by chance tested for speed?  frequently hand made for-loops in Julia are faster than the convenience functions.

Reply:
This does the same thing (while allowing for more types of numbers of course). I haven't benchmarked yet, but I'll do so.  Hadn't thought this method would have a large performance impact, but I see now that there's every potential it could since it loops through the entire color aesthetic at least once.

While I'm digging in, I'll see if it's possible to make this method type-stable.

Reply:
This does the same thing (while allowing for more types of numbers of course). I haven't benchmarked yet, but I'll do so. Hadn't thought this method would have a large performance impact, but I see now that there's every potential it could since it loops through the entire color aesthetic at least once.

While I'm digging in, I'll see if it's possible to make this method type-stable.

Reply:
Ran a little mini-diagnostic where I encapsulated the changed functionality into
a local function and called it. Example for the new version:

```julia
    function internal_test(datas::Gadfly.Data...)
        cdata = reduce(vcat, [], [i.color for i in datas if i.color != nothing])
        filter!(!isna, cdata)
        if !isempty(cdata)
          cmin, cmax = extrema(cdata)
        else
            return
        end
    end
    @time cmin, cmax = internal_test(datas...)
```

(Yes, yes I know it's not correct, but it's correct for the happy path, and
that's enough for this benchmark). Results for the new version:

```
julia> using Gadfly;
julia> x = 1:10000;
julia> y = rand(1:100, 10000);
julia> c = rand(1:250, 10000);
julia> plot(x=x, y=y, color=c)
  0.212279 seconds (93.11 k allocations: 5.061 MiB)
  0.000835 seconds (22 allocations: 78.953 KiB)
  0.000785 seconds (22 allocations: 78.953 KiB)
julia> plot(x=x, y=y, color=c)
  0.001313 seconds (22 allocations: 78.953 KiB)
  0.000730 seconds (22 allocations: 78.953 KiB)
  0.000760 seconds (22 allocations: 78.953 KiB)
```

Modified code for old version:

```julia
    function internal_test(datas::Gadfly.Data...)
        cmin, cmax = Inf, -Inf
        for data in datas
            data.color === nothing && continue

            for c in data.color
                c === NA && continue

                c = convert(Float64, c)
                if c < cmin
                    cmin = c
                end

                if c > cmax
                    cmax = c
                end
            end
        end

        (cmin == Inf || cmax == -Inf) && return
        cmin, cmax
    end
    @time cmin, cmax = internal_test(datas...)
```

Results for old version:

```
julia> using Gadfly;
julia> x = 1:10000;
julia> y = rand(1:100, 10000);
julia> c = rand(1:250, 10000);
julia> plot(x=x, y=y, color=c)
  0.023244 seconds (33.02 k allocations: 805.858 KiB)
  0.002329 seconds (29.50 k allocations: 617.172 KiB)
  0.002437 seconds (29.50 k allocations: 617.172 KiB)
julia> plot(x=x, y=y, color=c)
  0.002203 seconds (29.50 k allocations: 617.172 KiB)
  0.002188 seconds (29.50 k allocations: 617.172 KiB)
  0.002348 seconds (29.50 k allocations: 617.172 KiB)
```

Believe it or not, this change is actually a performance improvement if
anything. Looks like the compiler can basically optimize out the new stuff.